### ISSUE TITLE
tf.nn.bias_add does not support multiple derivatives (using with contrib.learn)

### ISSUE THREAD

**Issue by admcrae** [tensorflow/tensorflow#4174]
Apparently, `tf.nn.bias_add` does not support differentiating twice: I created a neural network with `tf.contrib.layers`, and attempting to compute a Hessian matrix (actually Hessian-vector product, based on https://en.wikipedia.org/wiki/Hessian_automatic_differentiation) yields a `LookupError` saying that `BiasAddGrad` has no gradient.

From what I understand, `bias_add` is basically `add` with support for different types.

Is there any way to allow `bias_add` to be differentiated multiple times? Perhaps at the very least we could make the use of `bias_add` optional in `layers`.

**Comment by tatatodd** [comment]
FYI a related issue is https://github.com/tensorflow/tensorflow/issues/2598

@rmlarsen can you comment on whether multiple derivatives of `bias_add` would be feasible?  Thanks!

**Comment by rmlarsen** [comment]
I think the best solution is to explicitly add a second derivative function in Python, similar to what I did for InvGrad here:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L255

@tatatodd would you be willing to send a PR for that?

**Comment by rmlarsen** [comment]
I meant @admcrae would you be willing to send a PR? @tatatodd is welcome to as well ;-)

**Comment by admcrae** [comment]
Yes, I can do it.

Out of curiousity (I'm extremely new to the inner workings of TensorFlow ops), why do ops like `BiasAddGrad` and `InvGrad` need C++ kernels?  It looks as though both could easily be implemented with more elementary operations (e.g. a reduction and possibly a casting in the former case, and a simple squaring + multiplication in the latter case). For that matter, why do `inv` and `div` need separate kernels at all? Is it all for performance optimization?

**Comment by rmlarsen** [comment]
@admcrae Sounds great! 

We added C++ kernels for them purely for performance, since they appear often large DL  models and/or their gradients. For example, div(scalar, tensor) would first broadcast scalar to a tensor then divide the two tensor elementwise, whereas inv(tensor) will apply the unary transformation x = 1/x elementswise. The actual division will of course used vectorized/SIMD instructions on most hardware.

**Comment by admcrae** [comment]
Just submitted a PR. It's a bit ugly (generating a zero tensor and applying bias_add to get the required broadcasting), so I'd be happy to hear if anyone has a better suggestion.

**Comment by admcrae** [comment]
I'll wait to close this until the PR gets merged.
